karapace/protobuf/proto_normalizations.py

```diff
@@ -0,0 +1,277 @@
+"""
+Copyright (c) 2024 Aiven Ltd
+See LICENSE for details
+"""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from karapace.dependency import Dependency
+from karapace.protobuf.enum_constant_element import EnumConstantElement
+from karapace.protobuf.enum_element import EnumElement
+from karapace.protobuf.extend_element import ExtendElement
+from karapace.protobuf.field_element import FieldElement
+from karapace.protobuf.group_element import GroupElement
+from karapace.protobuf.message_element import MessageElement
+from karapace.protobuf.one_of_element import OneOfElement
+from karapace.protobuf.option_element import OptionElement
+from karapace.protobuf.proto_file_element import ProtoFileElement
+from karapace.protobuf.rpc_element import RpcElement
+from karapace.protobuf.schema import ProtobufSchema
+from karapace.protobuf.service_element import ServiceElement
+from karapace.protobuf.type_element import TypeElement
+from karapace.schema_references import Reference
+from typing import Mapping, Sequence
+
+
+def sort_by_name(element: OptionElement) -> str:
+    return element.name
+
+
+class NormalizedRpcElement(RpcElement):
+    pass
+
+
+class NormalizedServiceElement(ServiceElement):
+    rpcs: Sequence[NormalizedRpcElement] | None = None
+
+
+class NormalizedFieldElement(FieldElement):
+    pass
+
+
+class NormalizedExtendElement(ExtendElement):
+    fields: Sequence[NormalizedFieldElement] | None = None
+
+
+class NormalizedTypeElement(TypeElement):
+    nested_types: Sequence[NormalizedTypeElement]
+
+    @abstractmethod
+    def to_schema(self) -> str:
+        pass
```

Comment:
```suggestion
class NormalizedTypeElement(TypeElement, abc.ABC):
    nested_types: Sequence[NormalizedTypeElement]
```

suggestion: If pylint complains, I'd just ignoring it. We should prefer marking the class final over the repeating the empty definition.

aside: In ABCs, it's a good idea to raise `NotImplementedError` in abstract definitions.